Purre ordering domain model

retinio opened the pull request:
Hi!
I suggest some improvements for ordering domain model.
It makes the model cleaner and  independent of the infrastructure (MediatR.dll)

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet-architecture/eShopOnContainers?pullRequest=516) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet-architecture/eShopOnContainers?pullRequest=516) before we can accept your contribution.<br/><br/>:x: retinio [sign now](https://cla.dotnetfoundation.org/dotnet-architecture/eShopOnContainers?pullRequest=516)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet-architecture/eShopOnContainers?pullRequest=516) it.</sub>

CESARDELATORRE commented:
@retinio - Thanks for your PR. We really appreciate it. 👍 
Before we review the code and since it is changing quite a few files (21), could you add a comment to each of the important changes you made explaining the reasons for each change?
It could also be a good idea to split the PR in multiple PRs if the changes are related to different features or different improvements. That's up to you. :)
Thanks,

retinio commented on src/Services/Ordering/Ordering.API/Application/DomainEventHandlers/BuyerAndPaymentMethodVerified/UpdateOrderWhenBuyerAndPaymentMethodVerifiedDomainEventHandler.cs:
```diff
@@ -5,11 +5,12 @@
 {
     public class UpdateOrderWhenBuyerAndPaymentMethodVerifiedDomainEventHandler 
-                   : INotificationHandler<BuyerAndPaymentMethodVerifiedDomainEvent>
+                   : INotificationHandler<DomainEventNotification<BuyerAndPaymentMethodVerifiedDomainEvent>>
```
Use decorator IDomainEvent

retinio replied:
Use decorator IDomainEvent

retinio replied:
Use decorator IDomainEvent

retinio commented on src/Services/Ordering/Ordering.API/Application/DomainEventHandlers/OrderGracePeriodConfirmed/OrderStatusChangedToAwaitingValidationDomainEventHandler.cs:
```diff
@@ -12,7 +14,7 @@
 
     public class OrderStatusChangedToAwaitingValidationDomainEventHandler
-                   : INotificationHandler<OrderStatusChangedToAwaitingValidationDomainEvent>
+                   : INotificationHandler<DomainEventNotification<OrderStatusChangedToAwaitingValidationDomainEvent>>
```
Use decorator IDomainEvent

retinio replied:
Use decorator IDomainEvent

retinio commented on src/Services/Ordering/Ordering.API/Application/DomainEventHandlers/OrderPaid/OrderStatusChangedToPaidDomainEventHandler.cs:
```diff
@@ -12,7 +14,7 @@
 
     public class OrderStatusChangedToPaidDomainEventHandler
-                   : INotificationHandler<OrderStatusChangedToPaidDomainEvent>
+                   : INotificationHandler<DomainEventNotification<OrderStatusChangedToPaidDomainEvent>>
```
Use decorator IDomainEvent

retinio commented on src/Services/Ordering/Ordering.API/Application/DomainEventHandlers/OrderStartedEvent/ValidateOrAddBuyerAggregateWhenOrderStartedDomainEventHandler.cs:
```diff
@@ -6,11 +6,12 @@
 {
     public class ValidateOrAddBuyerAggregateWhenOrderStartedDomainEventHandler 
-                        : INotificationHandler<OrderStartedDomainEvent>
+                        : INotificationHandler<DomainEventNotification<OrderStartedDomainEvent>>
```
Use decorator IDomainEvent

retinio replied:
Use decorator IDomainEvent

retinio commented on src/Services/Ordering/Ordering.API/Application/DomainEventHandlers/OrderStockConfirmed/OrderStatusChangedToStockConfirmedDomainEventHandler.cs:
```diff
@@ -11,7 +13,7 @@
 
     public class OrderStatusChangedToStockConfirmedDomainEventHandler
-                   : INotificationHandler<OrderStatusChangedToStockConfirmedDomainEvent>
+                   : INotificationHandler<DomainEventNotification<OrderStatusChangedToStockConfirmedDomainEvent>>
```
Use decorator IDomainEvent

retinio commented on src/Services/Ordering/Ordering.Domain/Events/BuyerPaymentMethodVerifiedDomainEvent.cs:
```diff
@@ -1,10 +1,10 @@
 {
     public class BuyerAndPaymentMethodVerifiedDomainEvent
-        : INotification
+        : IDomainEvent
```
Use IDomainEvent instead INotifaction.

retinio replied:
Use IDomainEvent instead INotifaction.

retinio commented on src/Services/Ordering/Ordering.Domain/Events/OrderStartedDomainEvent.cs:
```diff
@@ -1,15 +1,13 @@
     /// Event used when an order is created
     /// </summary>
-    public class OrderStartedDomainEvent : INotification
+    public class OrderStartedDomainEvent : IDomainEvent
```
Use IDomainEvent instead INotifaction.

retinio commented on src/Services/Ordering/Ordering.Domain/Events/OrderStatusChangedToAwaitingValidationDomainEvent.cs:
```diff
@@ -1,14 +1,15 @@
     /// </summary>
     public class OrderStatusChangedToAwaitingValidationDomainEvent
-         : INotification
+         : IDomainEvent
```
Use IDomainEvent instead INotifaction.

retinio commented on src/Services/Ordering/Ordering.Domain/Events/OrderStatusChangedToPaidDomainEvent.cs:
```diff
@@ -1,14 +1,15 @@
     /// </summary>
     public class OrderStatusChangedToPaidDomainEvent
-        : INotification
+        : IDomainEvent
```
Use IDomainEvent instead INotifaction.

retinio commented on src/Services/Ordering/Ordering.Domain/Events/OrderStatusChangedToStockConfirmedDomainEvent.cs:
```diff
@@ -1,12 +1,12 @@
     /// </summary>
     public class OrderStatusChangedToStockConfirmedDomainEvent
-        : INotification
+        : IDomainEvent
```
Use IDomainEvent instead INotifaction.

retinio commented on src/Services/Ordering/Ordering.Domain/Ordering.Domain.csproj:
```diff
@@ -5,8 +5,6 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="MediatR" Version="4.0.1" />
```
Remove infrastructure library (MediatR) from domain layer

retinio commented on src/Services/Ordering/Ordering.Domain/SeedWork/Entity.cs:
```diff
@@ -20,18 +21,22 @@ protected set
+        public IEnumerable<IDomainEvent> DomainEvents => _domainEvents?.AsReadOnly();       
         
-        public void AddDomainEvent(INotification eventItem)
+        public void AddDomainEvent(IDomainEvent eventItem)
```
Use IDomainEvent instead INotifaction.

retinio commented on src/Services/Ordering/Ordering.Infrastructure/MediatorExtension.cs:
```diff
@@ -19,14 +22,24 @@ static class MediatorExtension
             await Task.WhenAll(tasks);
         }
+
+        public static Task PublishEvent(this IMediator mediator, IDomainEvent domainEvent, CancellationToken cancellationToken)
```
Extension method PublishEvent was created for publishing domain event.

retinio commented:
@CESARDELATORRE  – Thanks for your feedback.
My PR consists of several parts of one whole.
1.	I have  added IDomainEvent and changed Entity class. Now Entity implements IDomainEvent instead INotifaction from MediatR. This refactoring allows removing infrastructure library (MediatR) from domain layer.
2.	Decorator DomainEventNotification<T> was created for using in each DomainEventHandler
3.	Extension method PublishEvent was created for publishing domain event.
I have added review comments  for each file changed. 
Thanks.

retinio commented on src/Services/Ordering/Ordering.API/Application/DomainEventHandlers/OrderPaid/OrderStatusChangedToPaidDomainEventHandler.cs:
```diff
@@ -27,16 +29,16 @@ public class OrderStatusChangedToPaidDomainEventHandler
         }
 
-        public async Task Handle(OrderStatusChangedToPaidDomainEvent orderStatusChangedToPaidDomainEvent, CancellationToken cancellationToken)
+        public async Task Handle(DomainEventNotification<OrderStatusChangedToPaidDomainEvent> orderStatusChangedToPaidDomainEventNotification, CancellationToken cancellationToken)
```
Use decorator IDomainEvent
